frontend/components/app.jsx

```diff
@@ -1,61 +1,46 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 import PropTypes from 'prop-types';
 import Notifications from 'react-notification-system-redux';
 import { connect } from 'react-redux';
+import { useLocation } from '@reach/router';
 
 import { ALERT } from '../propTypes';
 import alertActions from '../actions/alertActions';
 import BuildStatusNotifier from '../util/buildStatusNotifier';
 
-export class App extends React.Component {
-  componentDidMount() {
-    const { notifier, onEnter } = this.props;
-    onEnter();
-    notifier.listen();
+function shouldClearAlert(alert) {
+  if (alert.stale) {
+    alertActions.clear();
+    return;
   }
 
-  // eslint-disable-next-line camelcase
-  UNSAFE_componentWillReceiveProps(nextProps) {
-    const { alert } = this.props;
-
-    if (alert.message) {
-      this.shouldClearAlert(nextProps);
-    }
+  if (alert.message) {
+    alertActions.update(alert.stale);
   }
+}
+export function App(props) {
+  const {
+    alert, notifier, children, notifications, onEnter,
+  } = props;
+  const location = useLocation();
 
-  shouldClearAlert({ alert, location }) {
-    // the route we are leaving
-    const { location: { key: lastKey } } = this.props;
-    // the route we are moving to
-    const { key: nextKey } = location;
-
-    if (alert.stale) {
-      alertActions.clear();
-      return;
-    }
-
-    // clear an existing alert message if stale, or flag it to be removed on
-    // the next route transition
-    if (lastKey === nextKey) return;
+  useEffect(() => {
+    onEnter();
+    notifier.listen();
+  }, []);
 
+  useEffect(() => {
     if (alert.message) {
-      alertActions.update(alert.stale);
+      shouldClearAlert(alert);
     }
-  }
-
-  render() {
-    const {
-      children,
-      notifications,
-    } = this.props;
+  }, [location.key]);
```

Comment:
Do we need to also add the `alert` prop into the array of the `useEffect` argument so the effect will fire when the `alert` prop is updated?

Reply:
So I'm not positive about the logic here but I purposefully removed `alert` from the dependencies to simplify the logic. By my understanding, the previous implementation and intent was:
1. fire on every props change
2. if the alert is stale, clear it
3. if the location key matches, early return (no action)
4. if there is an existing message during a location key change, we should mark it as stale for removal on the next props update

The new logic should _mostly_ retain that because it will only fire on location.key changes (skips the early return) and will mark the message as stale + clear stale messages, the one difference is that it requires a location change to clear the message (which wasn't true before).

I wasn't sure of a great way to test/simulate this behavior though

Reply:
Thanks for laying out your thought process. It makes sense to me and let's see how it goes.